### ISSUE TITLE
v5: is it possible to save scope and restore? 

### ISSUE THREAD

**Issue by nohwnd** [pester/Pester#1227]
I figure if someone would know it would be you @alx9r :) 

Do you know if there is a way to save a scope, pop it, run some code, then push it back in and run some more code? 

I made some assumptions in Pester v5 that don't seem to hold. I was trying to make the `It` blocks isolated from describes, but did not realize the putting describes in between `It` blocks would inherit the state. This coupled with running the setup before first `It` and not before the `Describe` that holds it leads to a very confusing behavior. I am trying to explain it in more detail here: [v5-alpha2: nested blocks and their setups](https://github.com/pester/Pester/blob/v5.0/README.md#-nested-blocks-and-their-setups). 

```powershell
# so this
$a = 1 
beforeAll { $a  = 2 }
it { $a # a is 2 }
describe { $a # a = 1 }
it { $a # a = 2 }

# would execute like this

$a = 1
&{
    # scope for tests
    . { #before all
        $a  = 2 
    }
    & {
        # first it
        $a
    }
    # can I save this scope?
}
& {
    # describe, $a is 1 here
    $a 
}
&{
    # and restore it here so $a is 2 again?
    &{
        $a
    }
}
```

**Comment by SeeminglyScience** [comment]
> Do you know if there is a way to save a scope, pop it, run some code, then push it back in and run some more code?

It's very unlikely.  If it *is* possible, it would require mountains of reflection and would be incredibly fragile.

That said, you may be able to get a similar effect through other methods.

## Closures

The only way I can think of to get the *exact* flow you're going for here is using dynamic module closures.  Not sure how useful that's going to be here as it's definitely not side effect free.

```powershell
$a = 1
$itClosure = { }.GetNewClosure()

&{
    # scope for tests
    . $itClosure.Module { #before all
        $a = 2
    }
    & $itClosure.Module {
        # first it
        $a
    }
    # can I save this scope?
}
& {
    # describe, $a is 1 here
    $a
}
&{
    # and restore it here so $a is 2 again?
    & $itClosure.Module {
        $a
    }
}
```

That's essentially saving a scope as a session state.  But that changes what the `script` scope references, and likely a lot of other unintended side effects.

## Execution Order

The only way I know of to scope those the way you'd like is to change the order they are executed in.

Here's a rough mock up

```powershell
function Describe {
    [CmdletBinding()]
    param(
        [string] $Name,
        [scriptblock] $Body
    )
    begin {
        function CreateBlock {
            [CmdletBinding()]
            param(
                [string] $Name,
                [scriptblock] $Body
            )
            end {
                # yield
                [PSCustomObject]@{
                    BlockType = $MyInvocation.MyCommand.Name
                    Name = $Name
                    Body = $body
                }
            }
        }
    }
    end {
        $describeStack = [System.Collections.Generic.Stack[psobject]]::new()
        $describeStack.Push(
            [PSCustomObject]@{
                BlockType = 'Describe'
                Name = $Name
                Body = $Body
            })

        while ($describeStack.Count -gt 0) {
            $currentBody = $describeStack.Pop().Body
            $innerBlocks = $currentBody.InvokeWithContext(
                <# functionsToDefine: #> @{
                    Describe = $function:CreateBlock
                    It = $function:CreateBlock
                    BeforeAll = $function:CreateBlock
                    AfterAll = $function:CreateBlock
                },
                <# variablesToDefine: #> @(),
                <# args:              #> @())

            $beforeAll = $null
            $afterAll = $null
            $itBlocks = [System.Collections.Generic.List[psobject]]::new()

            foreach ($block in $innerBlocks) {
                switch ($block.BlockType) {
                    BeforeAll { $beforeAll = $block }
                    AfterAll { $afterAll = $block }
                    It { $itBlocks.Add($block) }
                    Describe { $describeStack.Push($block) }
                }
            }

            & {
                if ($beforeAll) {
                    . $beforeAll.Body
                }

                foreach ($itBlock in $itBlocks) {
                    & $itBlock.Body
                }

                if ($afterAll) {
                    . $afterAll.Body
                }
            }
        }
    }
}
```

And demo

```powershell
Describe test {
    $a = 1
    beforeAll a { $a  = 2 }
    it b { write-host b $a}
    describe c { write-host c $a }
    it d { write-host d $a }
}
```

Returns

```raw
b 2
d 2
c 1
```

The general idea is to save the blocks as objects first, then execute them in an order you choose.  Not sure if Pester already does something similar though.

**Comment by nohwnd** [comment]
@SeeminglyScience thanks a lot :) Running the blocks out of order is a possibility that I considered, but I think it makes the behavior too surprising. In v5 you are not supposed to put code in between the Pester controlled blocks, but you are still able to share state via reference or via side effects in external resources. Running out of order would make this harder to understand, while I don't even know if the functionality I proposed is useful.

> *TL;DR:* A rant where I mostly try to explain to myself how it should work follows 😁 

There are multiple options how to approach this now: 
- Run BeforeAll and AfterAll exactly where it is placed, (but still run BeforeEach and AfterEach before and after each test of course). 
- Run `BeforeAll` right after start of `Describe` that contains it. (this does not allow defining anything before `BeforeAll` by for example using `Add-Dependency` in that describe)
- Run `BeforeAll` before any `Describe` or `It`. (this allows defining something before `BeforeAll` by for example using `Add-Dependency` in that describe)
- Make the current `BeforeAll` to `BeforeAllIts` and add a new `BeforeAll` keep the current behavior and tell people to always put tests after any describe
- Make it even more complicated... 😁 

There is also a request for block that will run before each block, but not before each test...

So let's list that: 
- Run code before all describes
- Run code before every test
- Run code before every describe (but not test)
- Run code before all tests (but not describe) 

So given the limitation of how we can manipulate scopes this can be done as follows 

- Run code before all describes -> either run the code from BeforeAll after the start of Describe before first Describe/It. I'd go with the latter because potential stray code would run before the setup, giving it a flow that it easier to understand, not like in v4 where the code would execute even before the containing describe. (Another even less surprising option would be to just run the code where it is defined, but this would also mean that the block should be put into the correct place, and BeforeEach would behave differently, so let's not do that.)
- Run code before every test -> this is already done and simple, if the current block contains test setup I dot source it to every test.
- Run code before every describe (but not test), this is not in v4 but should be simple to add, the code should imho run right before the describe starts. I think this is already implemented but not published.
-  Run code before all tests (but not describe), this runs into the limitations higlighted above, and can be easily "emulated" by adding another describe around the tests without all the other weirdness - so I will keep this implemented but won't publish it right now.

To ilustrate this: 

```powershell
Describe {
    BeforeAll { } 
    BeforeEach { }
    BeforeEachBlock {  }
    Describe {
        $a = 10
        It {
        }
        It { 
        } 
    }
}
```
```
Describe {
    BeforeAll {   } 
    BeforeEach { }
    BeforeEachBlock {  }
    # before all runs here, because this is the first Describe / It in this block
    Describe {
        # before each block runs here, it runs in the sb to isolate setups of different blocks
        $a = 10
        It {
            # before each runs here, because it was inherited from parent
        }
        It { 
            # before each runs here, because it was inherited from parent
        } 
    }
}
```
This should imho work fine, and provide all the functionality needed. 
I omitted that the setups and teardowns accumulate, in a way that runs the setups that are defined deeper in the structure later, and teardowns that are deeper earlier.

**Comment by nohwnd** [comment]
@alx9r thanks, for looking into this, but no. it is an interesting concept, and I would love to see it work, but but I am afraid it would be too hard to debug (and explain) to put it into Pester.

—-

As for the test isolation, I haven’t done much measuring but I would expect that running new runspace per It would be too granular, and would prevent you from sharing one guard Mock per describe or a common setup. On the othe hand a new runspace per file, is something that would be really easy, have been experimented with and would keep the language working as is.

**Comment by SeeminglyScience** [comment]
One more idea.

You could manually capture and restore state.  There's no real way to capture *everything* using only public API's though.  Here's how you could get all the internal dictionaries attached to `SessionStateScope` (including defined PowerShell classes)

```powershell
using namespace System
using namespace System.Collections
using namespace System.Collections.Generic
using namespace System.Linq.Expressions
using namespace System.Management.Automation
using namespace System.Reflection

function CreateStateExportDelegate {
    $flags = [BindingFlags]'Instance, NonPublic'

    # Cache reflection info
    $internalProperty = [SessionState].GetProperty('Internal', $flags)
    $currentScopeProperty = $internalProperty.PropertyType.GetProperty('CurrentScope', $flags)
    $sessionStateScopeType = $currentScopeProperty.PropertyType
    $getVariableTable = $sessionStateScopeType.GetProperty('Variables', $flags)
    $getFunctionTable = $sessionStateScopeType.GetProperty('FunctionTable', $flags)
    $getCmdletTable = $sessionStateScopeType.GetProperty('CmdletTable', $flags)
    $getAliasTable = $sessionStateScopeType.GetMethod('GetAliases', $flags)
    $getTypeTable = $sessionStateScopeType.GetProperty('TypeTable', $flags)

    $stateTables = $getVariableTable, $getFunctionTable, $getCmdletTable, $getAliasTable, $getTypeTable

    $dictionaryCtor = [Dictionary[string, object]].GetConstructor([type[]][IEqualityComparer[string]])
    $dictionaryAdd = [Dictionary[string, object]].GetMethod('Add', [type[]]([string], [object]))
    $ordinalIgnoreCaseProperty = [StringComparer].GetProperty('OrdinalIgnoreCase')

    # Create delegate
    $sessionState = [Expression]::Parameter([SessionState], 'sessionState')
    $currentScope = [Expression]::Variable($sessionStateScopeType, 'currentScope')
    $result = [Expression]::Variable([Dictionary[string, object]], 'result')

    $bodyExpr = [List[Expression]]::new()

    # currentScope = sessionState.Internal.CurrentScope;
    $bodyExpr.Add(
        [Expression]::Assign(
            $currentScope,
            [Expression]::Property(
                [Expression]::Property($sessionState, $internalProperty),
                $currentScopeProperty)))

    # result = new Dictionary<string, object>(StringComparer.OrdinalIgnoreCase);
    $bodyExpr.Add(
        [Expression]::Assign(
            $result,
            [Expression]::New.Invoke(
                $dictionaryCtor,
                [Expression]::Property($null, $ordinalIgnoreCaseProperty))))

    # result.Add("Name", currentScope.TableName);
    foreach ($member in $stateTables) {
        if ($member -is [PropertyInfo]) {
            $memberName = $member.Name
            $getterMethod = $member.GetMethod
        } else {
            $memberName = 'AliasTable'
            $getterMethod = $member
        }

        $bodyExpr.Add(
            [Expression]::Call(
                $result,
                $dictionaryAdd,
                [Expression]::Constant($memberName),
                [Expression]::Call($currentScope, $getterMethod)))
    }

    # return result;
    $bodyExpr.Add($result)

    $lambda = [Expression]::Lambda(
        [Expression]::Block(
            [ParameterExpression[]]($currentScope, $result),
            $bodyExpr),
        [ParameterExpression[]]($sessionState))

    return $lambda.Compile()
}

$getStateDelegate = CreateStateExportDelegate

& {
    . {
        $testVar = $true
    }

    $script:exportedState = $getStateDelegate.Invoke($ExecutionContext.SessionState)
}

$exportedState.Variables['testVar']
# Name                           Value
# ----                           -----
# testVar                        True
```

(You could do it without creating a delegate but it would be pretty slow)

**Comment by SeeminglyScience** [comment]
Also forgot to mention, to *import* you would:

1. Export the current state
1. Use a previously exported state to modify the dictionaries exported in step 1

The dictionaries here are the same ones used by the engine internally.  They are *not* copies.  So editing them *should* essentially be altering state directly.

**Comment by nohwnd** [comment]
I need to play with that at home, it looks really cool. Very unsupported though :/ (but the whole framework lies on unsupported api, and it never changed so far :D )

**Comment by nohwnd** [comment]
Thanks for the info here, I ended up doing it in a way that does not need this kind of magic, to keep it easy to debug. :)
